Automattic/jetpack
# Display notice when user has unactivated product license keys

This PR displays a ""You have an inactive product. Activate now" notice in the dashboard if the user has unactivated product license keys available.

Note: This PR is a "re-do", based off https://github.com/Automattic/jetpack/pull/21426 which was closed because of complications with failing build checks. You can reference #21426 to view the original comments.

#### Changes proposed in this Pull Request:


- Register API route (/jetpack/v4/licensing/user/counts)  for fetching user license counts.
- Fetch user license counts and store them in initialState.
- Register API route (/jetpack/v4/licensing/user/activation-notice-dismiss)  for storing activation notice dismissal info.
- Create reducer, actions, and selectors for retrieving and updating user license counts and notice dismissal info in redux state.
- Persists the notice's custom dismissal info in wp_options.
- Displays the notice when the user acquires a new license, or when the user has an available license(s), but has dismissed the notice and it's been over 2 weeks without activating it.
- Removes notice after a license key is activated, or if the notice is dismissed.
- 

#### Does this pull request change what data or activity we track or use?


No

### Testing instructions:






- Spin up a site using this Jetpack branch. 
- Connect Jetpack to your WordPress.com user account.
- Enable the license-key input box UI on the "My plan" page by adding `add_filter( 'jetpack_licensing_ui_enabled', '__return_true' );` into Jetpack.
- Go to `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page)
- Verify you **do not see a notice** at the top that says, "You have an inactive product. Activate now." (Assuming you don't have any unattached user-licenses associated to your user account, which you shouldn't...yet..)
- Verify you can see the License Key input field (at the bottom of the "My Plan" section).

Now obtain a 'user' product license key:

- log into your wpcom sandbox, and in file `wp-content/mu-plugins/0-sandbox.php` add: `add_filter( 'jetpack_license_checkout_m2_enabled', '__return_true' );`. (Enabling this flag makes Jetpack checkout issue you a license when completing checkout).
- Open a new command-line window in your sandbox and `tail -f /tmp/php-errors`
- Point `public-api.wordpress.com` to your sandbox IP (in your hosts file).
- Make sure you are proxied and go to https://wordpress.com/checkout/jetpack/jetpack_scan (siteless checkout flow).
- Complete checkout (using credits is fine).
- View your error log, you should see the output: `Jetpack user license was created for subscription_id:17829922, belonging to user_id:78351100 and blog_id:198416894`, but with different id values corresponding to your user and site. And you'll also see, `License key is: jetpack_scan_monthly_oRhIEF4ckyVP6mCuHaD1vYnw5`, but with your unique key.
- Copy the license key from the error log output, and paste it somewhere for reference. We'll need this license key in a later step.
- un-sandbox `public-api.wordpress.com`.

**Now test that the notice appears now that you have been provisioned a detached license**
- Go to your Jetpack site and reload `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page).
- Verify that you see a notice at the top that says, **"You have an inactive product. Activate now."**

**Now test the notice dismiss behavior:**
**Prerequisite**: You'll need [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) installed into your browser for this.

- Open your browser developer tools and open Redux DevTools. (you may need to reload the page.) Inspect the state tree and locate `state.licensing.activationNoticeDismiss`
- Click the notification dismiss button (x), verify that the notice disappears. and the values for `last_detached_count` and `last_dismissed_time` are updated.
- Now, to verify that the notice reappears when a new detached license in acquired, in Redux DevTools, dispatch the action below: (replacing the value of `detached` with the value of <last_detached_count + 1>.

```
{
	type: 'JETPACK_LICENSING_USER_LICENSE_COUNTS_UPDATE',
	counts: {
		attached: 0,
		detached: <enter the value of `activationNoticeDismiss.last_detached_count` + 1(plus one) here>,
		revoked: 0,
		not_revoked: 0,
	}
}
```
- Verify the notice appears again after the above action is dispatched (when a **new** detached license is added).
- Now reload the page, reloading the persisted `activationNoticeDismissed` values in state. The notice should currently be dismissed from when you clicked dismiss(x) button before.
- Now we'll dispatch another action, changing the `last_dismissed_time` to over 2 weeks ago: (`2021-10-13T16:59:22+00:00`)
```
{
	type: 'JETPACK_LICENSING_ACTIVATION_NOTICE_DISMISS_UPDATE',
	dismissData: {
		last_detached_count: <the current value in state, do not change>,
		last_dismissed_time: '2021-10-13T16:59:22+00:00' // <--change to this>
	}
}
```
- Verify the notice reappears (because the `last_dismiss_date` is set to over 2 weeks ago).

**Now test the notice disappears when you activate(submit) a license key:**
- First log into the site's `wp-shell`: (If using jurassic.ninja, ssh into the site using the ssh user and pass that is displayed in the Dashboard, then run `wpsh`. If using local docker env, run `jetpack docker wp shell`)
- run command: `update_option('jetpack_licensing_activation_notice_dismiss', array('last_detached_count' => null, 'last_dismissed_time' => null) );` . This will reset the dismiss notice.
- Go to your Jetpack site and reload `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page). Verify the activation notice is visible.
- Go back to Calypso, https://wordpress.com/purchases/subscriptions/:site, and copy the license key you purchased in the beginning steps.
- Paste the license key into the License Key input box and submit.
- Verify the "You have an inactive product. Activate now." notice **disappears** (Also you should see that the Jetpack Scan product has been provisioned for your site(You may need to refresh the page to see that))


**Screenshot showing the new notice ("You have an inactive product. Activate now"):**

![Markup 2021-10-29 at 07 25 37](https://user-images.githubusercontent.com/11078128/139426809-34f1568a-9d9b-4d16-8be0-e3bea4136e3c.png)
